pkg/dockerregistry/server/writelimiter/blobstore.go

```diff
@@ -0,0 +1,107 @@
+package writelimiter
+
+import (
+	"runtime"
+	"sync/atomic"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+
+	"github.com/openshift/origin/pkg/dockerregistry/server/types"
+)
+
+type blobWriter struct {
+	distribution.BlobWriter
+	release func()
+}
+
+var _ distribution.BlobWriter = blobWriter{}
+
+func (bw blobWriter) Close() error {
+	bw.release()
+	return bw.BlobWriter.Close()
+}
+
+type blobStore struct {
+	distribution.BlobStore
+	sem chan struct{}
+}
+
+var _ distribution.BlobStore = blobStore{}
+
+func (bs blobStore) acquire() func() {
+	// LOCK_BEGIN++
+	sem := bs.sem
+	// LOCK_WAIT_BEGIN++
+	sem <- struct{}{}
+	// LOCK_WAIT_END++
+	done := int32(0)
+	return func() {
+		if atomic.SwapInt32(&done, 1) == 0 {
+			<-sem
+			sem = nil
```

Comment:
If you want to keep it, it is worth a comment. To me, it seems like the closure will be destroyed pretty soon when this returns. I'd expect the probability of gc freeing the `sem` before destroying the closure is pretty low.